### ISSUE TITLE
Limit the maximum parallelism on the DB level

### ISSUE THREAD

**Issue by vladzcloudius** [scylladb/scylla#7072]
## Description
Unlimited parallelism is a common issue we see a lot in the field.

When the DB gets hit by a burst of many outstanding requests that queue on a front end (Coordinator) this naturally leads to increased latencies and in extreme cases may cause the Scylla node to run out of memory (due to queuing of many requests contexts).

Fixing this today requires an application change which is many times not easy and sometimes impossible.

If we had the ability to control the maximum amount of queued up requests on the DB level without causing too much of service disturbance that would allow more time to developers to fix their application.

## Here is a proposal of how this may be implemented.
We can define two new parameters for each class of CQL requests we want to control (e.g. just for “Reads” and “Writes”):

 * `max_pending_XX` (int): maximum number of outstanding requests of the type XX for which we do not apply any back pressure.
 * `pending_XX_cutoff` (int): number of outstanding requests of the type XX after which each new request will be immediately reported as “Unavailable” without additional processing.

Then we can utilize the same flow control algorithm (code) we use for MVs today where we would “slow down” the response for every request XX between max_pending_XX and pending_XX_cutoff in the “XX foreground pending queue” and will respond with an “Unavailable” error to each new XX request that arrives if the outstanding (foreground) XX queue size is equal to pending_XX_cutoff.

The flow control code should do nothing as long as the size of the foreground queue is below `max_pending_XX` .

We should have this feature disabled by default (pending_XX_cutoff == 0).

We should have the ability to control this both globally and on the keyspace and table level.

**Comment by vladzcloudius** [comment]
@eliransin @slivne @avikivity @dorlaor FYI

**Comment by dorlaor** [comment]
@Piotr Sarna <sarna@scylladb.com> FYI

**Comment by psarna** [comment]
Note: while a total limit for concurrent statements is quite easy to check, separate limits for reads and writes would need to be checked considerably later - because the statement must first be parsed in order for Scylla to know if it's a read or write request. I initially thought that CQL protocol has separate opcodes for reads/writes but it's not the case. It also gets a little bit more complicated, because there are several code paths for statements to be executed (direct, prepared, internal, etc.) and multiple types of statements have to be properly grouped (e.g. SELECT should fall into read limits, while INSERT and UPDATE fall into write limits). Thus, Scylla will still need to spend some time and memory on parsing these extra packets, but at least they won't be executed, which is the most important thing. I'll try to create a PoC for that soon(ish).

**Comment by psarna** [comment]
Also, we currently have metrics for the total number of outstanding requests, but that counter is not split per request type (read/write). That would need to be changed as well, and these metrics need to be propagated to cql query processor, since that's where it's possible to see what kind of statement it is.

**Comment by avikivity** [comment]
In the general case, a single request can cause a large number of requests, and some of those can be performed in parallel. A request can map to very little load (CL_ONE from the same node/shard satisfied from cache) to a lot of load (indirection through an index, many sstable reads, each sstable read requiring a large partition index traversal, filtering away many tombstones and rows by WHERE clause). The more we evolve the language, the more complex a single query can become.

So counting CQL queries is not very meaningful.

**Comment by psarna** [comment]
I agree that in general this limit is not very meaningful, but if it can actually help some users recover from OOMs by setting a couple of parameters, then it may still make sense to add this. Under the assumption that a crude mechanism is still better than no mechanism.

**Comment by avikivity** [comment]
I don't think we can separate read and write. It somehow works now, but already we have requests that mix read and write (LWT). Also, if the same connection mixes read and write requests, and we hit (say) the read limit, then we won't read write requests either (head-of-line blocking).

We already have a solution for this in scylla-enterprise, workload prioritization, where we limit by workload instead of by read/write. I think we already limit query concurrency by estimating the amount of memory needed by a query and limiting the total amount of memory in use by transient requests.

**Comment by psarna** [comment]
> I don't think we can separate read and write. It somehow works now, but already we have requests that mix read and write (LWT). Also, if the same connection mixes read and write requests, and we hit (say) the read limit, then we won't read write requests either (head-of-line blocking).
> 
Then perhaps allowing to set a single concurrency limit for all types of requests makes some sense. It's also really easy to implement on top of current metrics. The use case I have in mind is where a cluster chronically fails because of concurrency spikes, so a field engineer sets a magic number of `50'000` requests tops and continue to investigate the root cause, while the cluster at least stops crashing. A PoC for that one can be coded really quickly.

> We already have a solution for this in scylla-enterprise, workload prioritization, where we limit by workload instead of by read/write. I think we already limit query concurrency by estimating the amount of memory needed by a query and limiting the total amount of memory in use by transient requests.

I think that the assumption for this issue is that the shedding based on memory estimation already failed (e.g. due to underestimating the memory consumption for reads), and we want to be extra safe against failures.

**Comment by avikivity** [comment]
It makes sense to have an upper limit (more or less equivalent to increasing the baseline memory consumption estimate).

Note that too little concurrency can cause availability failures. A write consumes a slot as long as we're missing an ack, and if a node is down, we'll run out of slots if it takes too long to detect it is down.

Similarly, it will limit throughput if we have cross-datacenter replication. Little says that C = L*T, so T=C/L. If C is lowered while L is high the throughput drop.s

**Comment by vladzcloudius** [comment]
> I don't think we can separate read and write. It somehow works now, but already we have requests that mix read and write (LWT). Also, if the same connection mixes read and write requests, and we hit (say) the read limit, then we won't read write requests either (head-of-line blocking).
> 
> We already have a solution for this in scylla-enterprise, workload prioritization, where we limit by workload instead of by read/write. I think we already limit query concurrency by estimating the amount of memory needed by a query and limiting the total amount of memory in use by transient requests.

@avikivity our current limits are only good for avoiding OOM situation and proven to be useless when you need to control the latency SLA.

This proposal is mostly about controlling the latency.

However controlling that we don't run out of memory should be done internally and is out of scope of this proposal.

As to your claim that counting CQL requests is not meaningful: each CQL requests queue up and based on the same Little law L=C/T and since the concurrency on the coordinator is C (because CL is usually fixed, TokenAware policy and prepared statements are used and tokens are nicely distributed) counting (and shedding) requests on the coordinator will allow us to control the latency in the majority of cases.

And since the CL may be assumed as constant - that will allow us control the parallelism on the replica as well.

On top of that I have already explained why shedding on the coordinator is much more efficient than shedding on the replica only.

I think we should shed both on the coordinator and on the replica however starting from the coordinator gives us much more benefit since it's going to address much more use cases.

@psarna I agree that separating reads and writes may be challenging - so let's start with something simple (account all requests) and then add per-request segregation later.

@avikivity Our workload prioritization doesn't address this at all because its arbitration is only active when there is a contention on resources. This proposal targets to operate at much lower levels of load.

**Comment by nyh** [comment]
@vladzcloudius @avikivity you started with discussing different read/write limits, but obviously there are far more absurd variations between different types of requests... The fact you can do 100,000 single-row CL=ONE reads per second, obviously doesn't say you can do 100,000 CREATE TABLE / DROP TABLE requests per second. We also have scans and various other queries that are much slower than the fastest requests. 
All in all, I think this feature should **only** be used in emergency situations to work around bugs - ordinary users should never set it, and it shouldn't be documented as something that ordinary users should consider doing. I think it should only be used in emergencies, for a very specific setup and workload which you are trying to "save". This should be made clear in the documentation.

**Comment by vladzcloudius** [comment]
> @vladzcloudius @avikivity you started with discussing different read/write limits, but obviously there are far more absurd variations between different types of requests... The fact you can do 100,000 single-row CL=ONE reads per second, obviously doesn't say you can do 100,000 CREATE TABLE / DROP TABLE requests per second. We also have scans and various other queries that are much slower than the fastest requests.
> All in all, I think this feature should **only** be used in emergency situations to work around bugs - ordinary users should never set it, and it shouldn't be documented as something that ordinary users should consider doing. I think it should only be used in emergencies, for a very specific setup and workload which you are trying to "save". This should be made clear in the documentation.

@nyh I think you are confusing this with the "rate limit" config (which this limit is NOT).

I wish we had a rate limiting capabilities (for background operations like compactions for instance) but this is something Asias is only planning to start working on.

**Comment by nyh** [comment]
> @nyh I think you are confusing this with the "rate limit" config (which this limit is NOT).

But it quacks like a duck, so how is it not a duck? I don't understand the difference. Sorry.

> I wish we had a rate limiting capabilities (for background operations like compactions for instance) but this is something Asias is only planning to start working on.

Background operations are yet another different story. I'm talking about very different CQL statements - compare "SELECT" of one row, to a "CREATE TABLE". Obviously no single concurrency limit will ever make sense for both of them.

Note that I'm not suggesting that we shouldn't have this option for emergencies. We just should make it very clear that this is not a configuration that ordinary users should tweak. Users shouldn't just see it in their scylla.yaml, and then start to think whether they want to set it to 10,000 or 100,000 or what. They should only use this option in case of emergency when the workload is already known - not while planning or sizing the cluster. I honestly don't see this option "widely used by many customers" , certainly not routinely - if this happens, I think we're doing something very wrong.

**Comment by vladzcloudius** [comment]
> > @nyh I think you are confusing this with the "rate limit" config (which this limit is NOT).
> 
> But it quacks like a duck, so how is it not a duck? I don't understand the difference. Sorry.

This PR is about limiting the max concurrency, not max rate.

If we take a mechanics approach to see the difference:
concurrency is measured in "requests"
rate is measured in "requests/sec".

I hope you see the difference now.

In particular we want to limit the maximum outstanding requests (currently of any kind) which is very different from limiting the maximum rate.

> 
> > I wish we had a rate limiting capabilities (for background operations like compactions for instance) but this is something Asias is only planning to start working on.
> 
> Background operations are yet another different story. I'm talking about very different CQL statements - compare "SELECT" of one row, to a "CREATE TABLE". Obviously no single concurrency limit will ever make sense for both of them.
> 
> Note that I'm not suggesting that we shouldn't have this option for emergencies. We just should make it very clear that this is not a configuration that ordinary users should tweak. Users shouldn't just see it in their scylla.yaml, and then start to think whether they want to set it to 10,000 or 100,000 or what. They should only use this option in case of emergency when the workload is already known - not while planning or sizing the cluster. I honestly don't see this option "widely used by many customers" , certainly not routinely - if this happens, I think we're doing something very wrong.

**Comment by nyh** [comment]
> This PR is about limiting the max concurrency, not max rate.

Yes, you are right, sorry (although the two things are closely related by the request latency (Little's Law)).
I am still worried that  a user who sees this option documented will start to over-think what to set there before he even has a problem... Should I set it to 10? to 1000? to 100,000? Are you planning to document how to choose this value?

**Comment by vladzcloudius** [comment]
It should be set to something like 30, @nyh (it's per shard).
This is our standard recommendation.

**Comment by nyh** [comment]
> It should be set to something like 30, @nyh (it's per shard).
> This is our standard recommendation.

Can such a low number be the recommendation? Just to give an example, this means that a 3-node, 8-shard-each cluster, will be limited to 720 concurrent clients. Try any more (because 800 of your clients happened to send a request around the same time), you get failures. Is this a good idea?

**Comment by vladzcloudius** [comment]
> > It should be set to something like 30, @nyh (it's per shard).
> > This is our standard recommendation.
> 
> Can such a low number be the recommendation? 

Yes it can.
You add more concurrency - your latency will suffer, @nyh 
The client can configure whatever they want eventually.
However when SLAs are defined around latencies, which is usually the case, limiting as above is not only not an issue - it's a requirement.

> Just to give an example, this means that a 3-node, 8-shard-each cluster, will be limited to 720 concurrent clients. Try any more (because 800 of your clients happened to send a request around the same time), you get failures. Is this a good idea?

Yes, it is - if you want to keep latencies at a certain level.
30 is just a base line. A user may try increasing it if more bandwidth is required.
However my c-s experiments shown that going beyond that concurrency doesn't give you more bandwidth because you start hitting the networking latency limit.

However for slower links one may try increasing it but, again, the CQL latency is going to suffer.
In practice, people never want to go beyond that.

**Comment by vladzcloudius** [comment]
Also keep in mind that those errors are usually transparent to the application thanks to a retry policy. If the app gets an overload error this means that the DB hints it that it's overloaded - which it is. In this case the app is supposed to try other nodes and/or apply an retry policy which usually has some back-off strategy (exponential is the most popular one).

There are certain corner cases in the current simplistic implementation, e.g. one client consumes all concurrency (queue) and the rest are starving. However when we introduce some "elasticity" to the concurrency queue like I suggested in my original proposal (#7072) then this should be addresses at some extent since we would hint the "faster" client that it should slow down.

However really malicious clients may still cause starvation similar to how a malicious TCP client may cause a starvation too.

However, again, the client should be really malicious to achieve that.
In a general case, if one client is too fast and others are slower - the too fast client will start getting more overloaded errors and, as a result, pace itself down allowing other clients to push in.
